handler/src/main/java/io/netty/handler/ssl/OpenSslCachingKeyMaterialProvider.java

```diff
@@ -0,0 +1,64 @@
+/*
+ * Copyright 2018 The Netty Project
+ *
+ * The Netty Project licenses this file to you under the Apache License,
+ * version 2.0 (the "License"); you may not use this file except in compliance
+ * with the License. You may obtain a copy of the License at:
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS, WITHOUT
+ * WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the
+ * License for the specific language governing permissions and limitations
+ * under the License.
+ */
+package io.netty.handler.ssl;
+
+import io.netty.buffer.ByteBufAllocator;
+
+import javax.net.ssl.X509KeyManager;
+import java.util.concurrent.ConcurrentHashMap;
+import java.util.concurrent.ConcurrentMap;
+
+/**
+ * {@link OpenSslKeyMaterialProvider} that will cache the {@link OpenSslKeyMaterial} to reduce the overhead
+ * of parsing the chain and the key for generation of the material.
+ */
+final class OpenSslCachingKeyMaterialProvider extends OpenSslKeyMaterialProvider {
+
+    private final ConcurrentMap<String, OpenSslKeyMaterial> cache = new ConcurrentHashMap<String, OpenSslKeyMaterial>();
```

Comment:
I'm not sure I understand exactly what key material is.  This looks like a potential memory leak, should I not be worried?

Reply:
@mosesn the `OpenSslKeyMaterial` is the combo of certificate chain and private key for an alias that is provided by the `X509KeyManager`. The important thing to note here is that the `OpenSslKeyMaterial` holds the pointers to the openssl structures and so native memory until these are destroyed. Usually you have 1 or two aliases. For example one for `rsa` and one for `ec` if you have dual chains or just one of these (which is usually the case). 

The `OpenSslKeyMaterial` and so the memory is released when `destroy()` is called, which is done once the `OpenSslContext` is destroyed. 

Does this make it more clear ?

Reply:
just to make it even clearer... this only caches what your X509KeyManager returns so it can not be "abused" by a remote peer.